## Q4. Standardization and Min-Max Scaling

---

### 🔍 এই Question-এ কী চাওয়া হয়েছে?

`monthly_income` column-এ দুটো scaling পদ্ধতি apply করতে হবে:

| Task | কী করতে হবে |
|---|---|
| **Q4.1** | Numeric column choose করা (`monthly_income`) |
| **Q4.2** | **Standardization (Z-score)** — mean=0, std=1 করা |
| **Q4.3** | **Min-Max Scaling** — [0, 1]-এ আনা |
| **Q4.4** | দুটোর পার্থক্য 2-3 বাক্যে comment করা |

### 🎯 এখান থেকে কী অর্জন করব?

দুটো scaling method-এর **output range ও behavior** পাশাপাশি দেখব — কোনটি কখন বেশি উপযুক্ত সেটার intuition তৈরি হবে।

### 🧠 দুটো Formula:

**Standardization (Z-score):**
$$z = \frac{x - \mu}{\sigma}$$
→ output range নির্দিষ্ট নয়, সাধারণত −3 থেকে +3-এর মধ্যে থাকে

**Min-Max Scaling:**
$$x' = \frac{x - x_{min}}{x_{max} - x_{min}}$$
→ output সবসময় **[0, 1]**-এ

### এই dataset-এর `monthly_income` stats:
- mean = **2885.75**, std = **898.12**
- min = **1000.00**, max = **5049.40**


### Q4.1 — Numeric Column Choose করা [2 Marks]

In [1]:
import pandas as pd
import numpy as np

# Now we'll load the dataset and select the numeric column
df = pd.read_csv("02. Provided - module123_data.csv")
num_col = "monthly_income"
# Here we've chosen monthly_income as our numeric column for scaling.

print("Selected column:", num_col)
print(df[num_col].describe().round(4))


Selected column: monthly_income
count     100.0000
mean     2885.7450
std       898.1247
min      1000.0000
25%      2379.1650
50%      2894.0200
75%      3593.1650
max      5049.4000
Name: monthly_income, dtype: float64


`monthly_income` column select করা হয়েছে।
`.describe()` দিয়ে mean=2885.75, std=898.12, min=1000, max=5049.40 দেখা যাচ্ছে।


### Q4.2 — Standardization with Z-score [10 Marks]

In [2]:
# Now we'll compute mean and standard deviation of the column
col_mean = df[num_col].mean()
col_std  = df[num_col].std()
# Here we've calculated mean=2885.75 and std=898.12 from the data.

print(f"Mean : {col_mean:.4f}")
print(f"Std  : {col_std:.4f}")


Mean : 2885.7450
Std  : 898.1247


`df[num_col].mean()` → সব মানের গড়।
`df[num_col].std()` → standard deviation — data কতটা ছড়িয়ে আছে।


In [3]:
# Now we'll apply the Z-score formula manually: z = (x - mean) / std
df["income_std"] = (df[num_col] - col_mean) / col_std
# Here we've created a new column 'income_std' with standardized values.

print("First 5 standardized values:")
print(df["income_std"].head().round(4).to_string())
print()
print(f"Standardized range: {df['income_std'].min():.4f}  to  {df['income_std'].max():.4f}")
print(f"Mean after scaling : {df['income_std'].mean():.6f}  (should be ~0)")
print(f"Std  after scaling : {df['income_std'].std():.6f}   (should be ~1)")


First 5 standardized values:
0    0.9447
1   -0.3246
2    0.7401
3    1.0415
4   -1.2636

Standardized range: -2.0996  to  2.4091
Mean after scaling : -0.000000  (should be ~0)
Std  after scaling : 1.000000   (should be ~1)


`(df[num_col] - col_mean) / col_std` → Z-score formula — mean বাদ দিয়ে std দিয়ে ভাগ।
**Output range:** −2.10 থেকে +2.41 — সব মান mean-এর কাছাকাছি, কোনো outlier নেই।
Scaling-এর পরে mean ≈ 0 এবং std ≈ 1 — এটাই Standardization-এর property।


### Q4.3 — Min-Max Scaling [10 Marks]

In [4]:
# Now we'll compute min and max of the column
col_min = df[num_col].min()
col_max = df[num_col].max()
# Here we've extracted min=1000.0 and max=5049.4 from the column.

print(f"Min : {col_min:.4f}")
print(f"Max : {col_max:.4f}")
print(f"Range: {col_max - col_min:.4f}")


Min : 1000.0000
Max : 5049.4000
Range: 4049.4000


`df[num_col].min()` ও `.max()` → scaling-এ ব্যবহৃত সর্বনিম্ন ও সর্বোচ্চ মান।


In [5]:
# Now we'll apply Min-Max formula manually: x' = (x - min) / (max - min)
df["income_minmax"] = (df[num_col] - col_min) / (col_max - col_min)
# Here we've created 'income_minmax' column with values scaled to [0, 1].

print("First 5 Min-Max scaled values:")
print(df["income_minmax"].head().round(4).to_string())
print()
print(f"Min-Max range: {df['income_minmax'].min():.4f}  to  {df['income_minmax'].max():.4f}")


First 5 Min-Max scaled values:
0    0.6752
1    0.3937
2    0.6298
3    0.6967
4    0.1854

Min-Max range: 0.0000  to  1.0000


`(df[num_col] - col_min) / (col_max - col_min)` → Min-Max formula।
**Output range:** 0.0 থেকে 1.0 — সবচেয়ে ছোট মান = 0, সবচেয়ে বড় মান = 1।


### দুটো Scaling পাশাপাশি দেখা

In [6]:
# Now we'll display original and both scaled columns side by side
comparison = df[[num_col, "income_std", "income_minmax"]].head(100)
# Here we've assembled the original and both scaled columns for easy comparison.

print("Comparison (first 10 rows):")
print(comparison.head().round(4).to_string(index=False))


Comparison (first 10 rows):
 monthly_income  income_std  income_minmax
        3734.19      0.9447         0.6752
        2594.19     -0.3246         0.3937
        3550.47      0.7401         0.6298
        3821.18      1.0415         0.6967
        1750.84     -1.2636         0.1854


Table-এ তিনটি column পাশাপাশি:
- **monthly_income** → original (1000–5049 range)
- **income_std** → Z-score (−2.10 থেকে +2.41)
- **income_minmax** → Min-Max (0.0 থেকে 1.0)


### Q4.4 — Short Answer [3 Marks]

> **Question:** Compare the standardized and min max scaled columns in 2 to 3 sentences. Mention what kind of range each one uses and how the numbers look.

---

**Answer:**

> The standardized column (`income_std`) centers the data around zero with a standard deviation of one, producing values that range from approximately −2.10 to +2.41, where negative values indicate incomes below the mean and positive values indicate incomes above it. The min-max scaled column (`income_minmax`) compresses all values into a fixed range of 0 to 1, where the lowest income maps to exactly 0 and the highest to exactly 1, making it easier to interpret as a proportion of the full range. While both methods remove the original unit (BDT), standardization is more suitable when the data is roughly normally distributed or when outliers should not distort the scale, whereas min-max scaling is preferred when a bounded output range is needed.

> Standardized column (`income_std`) data-কে শূন্যকে কেন্দ্র করে সাজায় এবং standard deviation হয় এক, ফলে মানগুলো প্রায় −২.১০ থেকে +২.৪১ পর্যন্ত বিস্তৃত হয়, যেখানে negative মান মানে গড়ের নিচের income আর positive মান মানে গড়ের উপরের income। Min-max scaled column (`income_minmax`) সব মানকে ০ থেকে ১-এর মধ্যে সংকুচিত করে, যেখানে সবচেয়ে কম income ঠিক ০-তে এবং সবচেয়ে বেশি income ঠিক ১-এ map করে — এতে পুরো range-এর অনুপাত হিসেবে বোঝা সহজ হয়। দুটি পদ্ধতিই মূল unit (BDT) সরিয়ে দেয়, তবে data যদি মোটামুটি normally distributed হয় বা outlier যেন scale বিকৃত না করে সেক্ষেত্রে standardization বেশি উপযুক্ত, আর নির্দিষ্ট সীমার মধ্যে output দরকার হলে min-max scaling-ই পছন্দনীয়।



---